test/drbgtest.c

```diff
@@ -470,29 +474,36 @@ static int error_check(DRBG_SELFTEST_DATA *td)
 
     /* Test explicit reseed with entropy source failure */
     t.entropylen = 0;
-    if (!TEST_int_le(RAND_DRBG_reseed(drbg, td->adin, td->adinlen, 0), 0)
+    if (!TEST_false(RAND_DRBG_reseed(drbg, td->adin, td->adinlen, 0))
             || !uninstantiate(drbg))
         goto err;
 
     /* Test explicit reseed with too much entropy */
     if (!instantiate(drbg, td, &t))
         goto err;
     t.entropylen = drbg->max_entropylen + 1;
-    if (!TEST_int_le(RAND_DRBG_reseed(drbg, td->adin, td->adinlen, 0), 0)
+    if (!TEST_false(RAND_DRBG_reseed(drbg, td->adin, td->adinlen, 0))
             || !uninstantiate(drbg))
         goto err;
 
     /* Test explicit reseed with too little entropy */
     if (!instantiate(drbg, td, &t))
         goto err;
     t.entropylen = drbg->min_entropylen - 1;
-    if (!TEST_int_le(RAND_DRBG_reseed(drbg, td->adin, td->adinlen, 0), 0)
+    if (!TEST_false(RAND_DRBG_reseed(drbg, td->adin, td->adinlen, 0))
             || !uninstantiate(drbg))
         goto err;
 
     /* Standard says we have to check uninstantiate really zeroes */
+    /*
+     * Nice idea, however RAND_DRBG_uninstantiate() cleanses the data via
+     * drbg_ctr_uninstantiate(), but right after that it resets drbg->data.ctr
+     * using RAND_DRBG_set(), so the following memcmp will fail.
+     */
+#if 0
     if (!TEST_mem_eq(zero, sizeof(drbg->data), &drbg->data, sizeof(drbg->data)))
         goto err;
+#endif
```

Comment:
Interesting. Note that there is a similar test in providers/fips/self_test_kats.c, which was added only recently by me in #11111. To make it work, I delayed the reinitialization of the drbg->data until the next instantiation. I didn't know that this was tested here, too.

Meanwhile this test should succeed, since your branch already contains my fixes:

* e70452155e Check that the DRBG's internal state has been zeroized after uninstantiation
* 75ff4f7404 DRBG: delay initialization of DRBG method until instantiation

So please revert this change.

Reply:
Some part of your comment could remain, to explain the check.

Reply:
You're right. It works on git master. My code was originally written for 1.1.1d, where the test failed. I'll re-enable the check, which should also fix the Travis build broken due to the unused variable zero.

Reply:
You are right, I was already thinking about that. But I did not have the time to write a reply yet.

We could indeed cherry-pick 75ff4f7404fe53304ad28279923760b61d6c2ace to 1.1.1 as a bug fix, since it breaks this test. @paulidale @slontis what's your opinion?

If this is not possible, we would have to  disable or remove the test in 1.1.1. It might be necessary to craft a separate pull request for 1.1.1. But for the moment, let's concentrate on master.

Reply:
I'm inclined to not cherry-pick the delayed reinitialisation.  1.1.1 has no need to formally verify that the internal state has been zeroed and there isn't a security problem since we know it will have been.

Reply:
Ok, then we'll just drop the test when cherry-picking to 1.1.1.